driver-core/src/main/com/mongodb/operation/MixedBulkWriteOperation.java

```diff
@@ -211,423 +226,443 @@ public void call(final AsyncConnection connection, final Throwable t) {
         });
     }
 
-    private void executeRunsAsync(final Iterator<Run> runs, final AsyncWriteBinding binding, final AsyncConnection connection,
-                                  final BulkWriteBatchCombiner bulkWriteBatchCombiner,
-                                  final SingleResultCallback<BulkWriteResult> callback) {
+    private BulkWriteResult executeBatches(final Connection connection, final SessionContext sessionContext,
+                                           final List<WriteRequestWithIndex> writeRequestsWithIndices) {
+        BulkWriteBatchCombiner bulkWriteBatchCombiner = new BulkWriteBatchCombiner(connection.getDescription().getServerAddress(),
+                ordered, writeConcern);
 
-        final Run run = runs.next();
-        final SingleResultCallback<BulkWriteResult> wrappedCallback = releasingCallback(callback, connection);
-        run.executeAsync(connection, new SingleResultCallback<BulkWriteResult>() {
-            @Override
-            public void onResult(final BulkWriteResult result, final Throwable t) {
-                if (t != null) {
-                    if (t instanceof MongoBulkWriteException) {
-                        bulkWriteBatchCombiner.addErrorResult((MongoBulkWriteException) t, run.indexMap);
-                    } else {
-                        wrappedCallback.onResult(null, t);
-                        return;
-                    }
-                } else if (result.wasAcknowledged()) {
-                    bulkWriteBatchCombiner.addResult(result, run.indexMap);
-                }
+        BatchMetadata batchMetadata = new BatchMetadata(writeRequestsWithIndices, 0);
+        do {
+            RawBsonDocument command = createCommand(connection.getDescription(), batchMetadata);
```

Comment:
Not a request for a change yet, but just wanted to note that that is going to double-copy the buffer: one copy for the RawBsonDocument, and then another for the wire protocol message.  

In scope of https://jira.mongodb.org/browse/JAVA-2584 this will have to change anyway, as the driver will have to split out the payload for the list of write requests in each write command.  We could either create the abstraction now that will allow for that, or wait until we work on that ticket, but I think it will look something like:

    interface Connection {
           <T> T command(String database, 
                                        BsonDocument command, 
                                        String payloadName,  // e.g. "documents" for insert command
                                        SomeAbstractionAroundBsonArray payload,    // e.g. the list of documents to insert
                                        ReadPreference readPreference, 
                                        FieldNameValidator fieldNameValidator,
                                        Decoder<T> commandResultDecoder, 
                                        SessionContext sessionContext);

SomeAbstractionAroundBsonArray would be a class that wraps a BsonArray but can also track how many of its members have been consumed.

SimpleCommandMessage would then have to handle the payload, either adding it to the command, in the case of OP_QUERY, or adding a second payload, in the case of OP_MSG.

Reply:
Do you prefer to handle in this PR or in a subsequent one?